lightning/src/ln/peers/handshake/mod.rs

```diff
@@ -0,0 +1,387 @@
+//! Execute handshakes for peer-to-peer connection establishment.
+//! Handshake states can be advanced automatically, or by manually calling the appropriate step.
+//! Once complete, returns an instance of Conduit.
+
+use secp256k1;
+
+use bitcoin_hashes::{Hash, HashEngine};
+use bitcoin_hashes::sha256::Hash as Sha256;
+use secp256k1::{PublicKey, SecretKey};
+
+use ln::peers::{chacha, hkdf};
+use ln::peers::conduit::Conduit;
+use ln::peers::handshake::acts::{ActOne, ActThree, ActTwo};
+use ln::peers::handshake::hash::HandshakeHash;
+use ln::peers::handshake::states::{ActOneExpectation, ActThreeExpectation, ActTwoExpectation, HandshakeState};
+
+pub(crate) mod acts;
+mod hash;
+mod states;
+mod tests;
+
+/// Object for managing handshakes.
+/// Currently requires explicit ephemeral private key specification.
+pub struct PeerHandshake {
+	state: Option<HandshakeState>,
+	private_key: SecretKey,
+	ephemeral_private_key: SecretKey,
+
+	read_buffer: Vec<u8>,
+}
+
+impl PeerHandshake {
+	/// Instantiate a new handshake with a node identity secret key and an ephemeral private key
+	pub fn new(private_key: &SecretKey, ephemeral_private_key: &SecretKey) -> Self {
+		let handshake = PeerHandshake {
+			state: Some(HandshakeState::Blank),
+			private_key: (*private_key).clone(),
+			ephemeral_private_key: (*ephemeral_private_key).clone(),
+			read_buffer: Vec::new(),
+		};
+		handshake
+	}
+
+	/// Make the handshake object inbound in anticipation of a peer's first handshake act
+	pub fn make_inbound(&mut self) {
+		let public_key = Self::private_key_to_public_key(&self.private_key);
+		let (hash, chaining_key) = Self::initialize_state(&public_key);
+		self.state = Some(HandshakeState::AwaitingActOne(ActOneExpectation {
+			hash,
+			chaining_key,
+		}))
+	}
+
+	fn initialize_state(public_key: &PublicKey) -> (HandshakeHash, [u8; 32]) {
+		// do the proper initialization
+		let protocol_name = b"Noise_XK_secp256k1_ChaChaPoly_SHA256";
+		let prologue = b"lightning";
+
+		let mut sha = Sha256::engine();
+		sha.input(protocol_name);
+		let chaining_key = Sha256::from_engine(sha).into_inner();
+
+		let mut initial_hash_preimage = chaining_key.to_vec();
+		initial_hash_preimage.extend_from_slice(prologue.as_ref());
+
+		let mut hash = HandshakeHash::new(initial_hash_preimage.as_slice());
+		hash.update(&public_key.serialize());
+
+		(hash, chaining_key) // hash, chaining_key
+	}
+
+	/// Process act dynamically
+	/// The role must be set before this method can be called
+	pub fn process_act(&mut self, input: &[u8], remote_public_key: Option<&PublicKey>) -> Result<(Vec<u8>, Option<Conduit>, Option<PublicKey>), String> {
```

Comment:
Some idle thoughts on this interface (`process_act`) and that of `process_act_two` and `process_act_three`:

Rather than returning tuples of `Option`s in `process_act` and the non-`Option` values in the other two methods, could they (i.e., `PublicKey` and `Conduit`) instead be set as fields in the appropriate act states (without using `Option`)? Then accessor methods on `PeerHandshake` could be provided for retrieving them (using `Option`). That way the code is a bit more self-documented and the return values of the act processing methods would be simpler (i.e., just returning the next Act to send).

Reply:
No, the act states are only what's necessary for the next step of the handshake. I think the handshake can be thought of as a chemical reaction, where there is an educt, a product, as well as some external inputs and byproducts. The handshake states are educts and products, and the conduit is a byproduct.

Reply:
What are the acts and the remote public key in this analogy?

Reply:
Discussed a bit offline. One idea is to have `PeerHandshake` initialized with the optional remote public key, so that it is not both an input and an output of `process_act`. And then have an accessor for accessing it. Similarly with the conduit.

Then the only return value would be the act data which could even be the `Act` enum instead given it simply wraps the vector, which would make the interface more explicit.